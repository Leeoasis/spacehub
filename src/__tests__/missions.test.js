import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import '@testing-library/jest-dom/extend-expect';
import MyMissions from '../components/missions';

const mockStore = configureMockStore();

describe('MyMissions component', () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      mission: [
        {
          mission_id: 'mission1',
          mission_name: 'Mission 1',
          description: 'Description 1',
          reserved: true,
        },
        {
          mission_id: 'mission2',
          mission_name: 'Mission 2',
          description: 'Description 2',
          reserved: false,
        },
      ],
    });
  });

  it('should render correctly', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <MyMissions />
      </Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should display missions', () => {
    const { getByText } = render(
      <Provider store={store}>
        <MyMissions />
      </Provider>,
    );
    expect(getByText('Mission 1')).toBeInTheDocument();
    expect(getByText('Mission 2')).toBeInTheDocument();
  });

  it('should show correct status for missions', () => {
    const { getByText } = render(
      <Provider store={store}>
        <MyMissions />
      </Provider>,
    );
    expect(getByText('Active Member')).toBeInTheDocument();
    expect(getByText('NOT A MEMBER')).toBeInTheDocument();
  });

  it('should dispatch joinMission action when join button is clicked', () => {
    const { getByText } = render(
      <Provider store={store}>
        <MyMissions />
      </Provider>,
    );
    fireEvent.click(getByText('Join Mission'));
    expect(store.getActions()).toEqual([{
      type: 'space-travelers-hub/missions/JOIN_MISSION',
      payload: {
        mission_id: 'mission2', mission_name: 'Mission 2', description: 'Description 2', reserved: false,
      },
    }]);
  });

  it('should dispatch leaveMission action when leave button is clicked', () => {
    const { getByText } = render(
      <Provider store={store}>
        <MyMissions />
      </Provider>,
    );
    fireEvent.click(getByText('Leave Mission'));
    expect(store.getActions()).toEqual([
      {
        type: 'space-travelers-hub/missions/LEAVE_MISSION',
        payload: {
          mission_id: 'mission1',
          mission_name: 'Mission 1',
          description: 'Description 1',
          reserved: true, // add reserved property
        },
      },
    ]);
  });
});
