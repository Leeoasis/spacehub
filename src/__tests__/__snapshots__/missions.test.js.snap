// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyMissions component should render correctly 1`] = `
<DocumentFragment>
  <div
    class="container mt -3"
  >
    <table
      class="table table-striped table-bordered table-hover"
    >
      <thead>
        <tr>
          <th>
            Mission
          </th>
          <th>
            Description
          </th>
          <th>
            Status
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th
            class="col-2"
          >
            Mission 1
          </th>
          <td
            class="col-6"
          >
            Description 1
          </td>
          <td
            class="col-2 text-center pt-5"
          >
            <span
              class=" active-btn bg-info text-white rounded-1 p-1"
            >
              Active Member
            </span>
          </td>
          <td
            class="col-2 text-center pt-5"
          >
            <button
              class="btn btn-light border-2 border-danger"
              name="leaveMission"
              type="submit"
            >
              Leave Mission
            </button>
          </td>
        </tr>
        <tr>
          <th
            class="col-2"
          >
            Mission 2
          </th>
          <td
            class="col-6"
          >
            Description 2
          </td>
          <td
            class="col-2 text-center pt-5"
          >
            <span
              class="bg-secondary text-white rounded-1 p-1"
            >
              NOT A MEMBER
            </span>
          </td>
          <td
            class="col-2 text-center pt-5"
          >
            <button
              class="btn btn-light border-2 border-secondary"
              type="submit"
            >
              Join Mission
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
